src/observer.jl

```diff
@@ -1,3 +1,237 @@
+"""
+CircuitObserver is an implementation of an
+observer object (<:AbstractObserver) which
+implements custom measurements to perform
+at each layer of the circuit evolution.
+"""
+struct CircuitObserver <: AbstractObserver
+  measurements::Dict{String, Union{Function, String, Tuple}}
+  results::Dict{String, Any}
+end
+
+"""
+    ITensors.measurements(observer::CircuitObserver, metric::String)
+
+Retrieve the results of a set of measurements
+"""
+ITensors.measurements(observer::CircuitObserver, metric::String) = observer.results[metric]
+
+
+CircuitObserver() = CircuitObserver(Dict{String, Union{Function, String, Tuple}}(),Dict{String, Function}())
+"""
+    CircuitObserver(observables::Dict{String, <:Any})
+
+Generate an observer given a list of measurements passed as a dictionary. 
+"""
+function CircuitObserver(measurements::Dict{String, <:Any}) 
+  res = Dict{String, Any}()
+  for obs in keys(measurements)
+    res[obs] = []
+  end
+  return CircuitObserver(measurements, res)
+end
+
+CircuitObserver(measurement::Pair{String, <:Any}) = 
+  CircuitObserver([measurement])
+
+CircuitObserver(measurements::Vector{<:Pair{String, <:Any}}) = 
+  CircuitObserver(Dict(measurements))
+
+
+function CircuitObserver(measurements::Vector{<:Any})
+  res = Dict{String, Any}()
+  named_measurements = Dict{String, Union{Function, String, Tuple}}()
+  for measurement in measurements
+    name = measurement_name(measurement)
+    named_measurements[name] = (measurement isa Pair ? last(measurement) : measurement)
+    res[name] = []
+  end
+  return CircuitObserver(named_measurements, res)
+end
+
+CircuitObserver(measurement::Union{String,Tuple,Function}) = 
+  CircuitObserver([measurement])
+
+
+
+function measurement_name(measurement::Union{String,<:Tuple})
+  name = (measurement isa String ? measurement*"(n)" : "")
```

Comment:
Can we just leave out the `"(n)"`? 

Also, this seems like a good place to use dispatch, i.e. split it up into:
```julia
measurement_name(measurement::String) = measurement
function measurement_name(measurement::Tuple)
  [...]
end
```
I'm also not a huge fan of how the current printing looks. Maybe we could do:
```julia
("X", 2) -> "X(2)"
("X", 1:4) -> "X(1:4)"
("X", "Y") -> "XY"
("X", 2, "Y", 3) -> "X(2)Y(3)"
```
It should be easy enough to write, basically anything that is not a string gets wrapped in parentheses:
```julia
add_parens(x::Tuple) = ntuple(n -> x[n] isa AbstractString ? x[n] : "("*string(x[n])*")", length(x))
measurement_name(measurement::Tuple) = prod(add_parens(measurement))
```